builtin/common/serialize.lua

```diff
@@ -1,205 +1,101 @@
+					elseif n_tp == "boolean" then
+						n_key = k and "true" or "false"
+					elseif n_tp == "nil" then
+						n_key = "nil"
+					elseif n_tp == "number" then
+						n_key = string_format("%.17g", k)
+					elseif n_tp == "string" then
+						n_key = string_format("%q", k)
+					else
+						error("Invalid Datatype Used As Key")
+					end
+
+					local n = string_format("%s[%s]", name, n_key)
+					buff[#buff + 1] = n .. "=" .. stringify(n, v)
 				end
-			end
-			nested[x] = nil
-		end
-	end
 
-	local dumped     = {}  -- object->varname set
-	local local_defs = {}  -- Dumped local definitions as source code lines
-
-	-- Mutually recursive local functions:
-	local dump_val, dump_or_ref_val
-
-	-- If x occurs multiple times, dump the local variable rather than
-	-- the value. If it's the first time it's dumped, also dump the
-	-- content in local_defs.
-	function dump_or_ref_val(x)
-		if seen[x] ~= 2 then
-			return dump_val(x)
-		end
-		local var = dumped[x]
-		if var then  -- Already referenced
-			return var
-		end
-		-- First occurence, create and register reference
-		local val = dump_val(x)
-		local i = local_index
-		local_index = local_index + 1
-		var = "_["..i.."]"
-		local_defs[#local_defs + 1] = var.." = "..val
-		dumped[x] = var
-		return var
-	end
-
-	-- Second phase.  Dump the object; subparts occuring multiple times
-	-- are dumped in local variables which can be referenced multiple
-	-- times.  Care is taken to dump local vars in a sensible order.
-	function dump_val(x)
-		local  tp = type(x)
-		if     x  == nil        then return "nil"
-		elseif tp == "string"   then return string.format("%q", x)
-		elseif tp == "boolean"  then return x and "true" or "false"
-		elseif tp == "function" then
-			return string.format("loadstring(%q)", string.dump(x))
-		elseif tp == "number"   then
-			-- Serialize numbers reversibly with string.format
-			return string.format("%.17g", x)
-		elseif tp == "table" then
-			local vals = {}
-			local idx_dumped = {}
-			local np = nest_points[x]
-			for i, v in ipairs(x) do
-				if not np or not np[i] then
-					vals[#vals + 1] = dump_or_ref_val(v)
-				end
-				idx_dumped[i] = true
+				return "{}\n" .. table_concat(buff, "\n")
+			else
+				return saved[value]
 			end
-			for k, v in pairs(x) do
-				if (not np or not np[k]) and
-						not idx_dumped[k] then
-					vals[#vals + 1] = "["..dump_or_ref_val(k).."] = "
-						..dump_or_ref_val(v)
-				end
-			end
-			return "{"..table.concat(vals, ", ").."}"
+		elseif tp == "function" then
+			return string_format("loadstring(%q)", string_dump(value, true))
+		elseif tp == "boolean" then
+			return value and "true" or "false"
+		elseif tp == "nil" then
+			return "nil"
+		elseif tp == "number" then
+			return string_format("%.17g", value)
+		elseif tp == "string" then
+			return string_format("%q", value)
 		else
-			error("Can't serialize data of type "..tp)
+			error("Invalid Datatype Passed To minetest.serialize()")
```

Comment:
Previous error message is preferrable, providing type info.

Reply:
Also Strings Like That Look Bad.